OSPSuite Architecture.md

```diff
@@ -0,0 +1,280 @@
+-	NHibernate (currently version 5.2.5). The ORM is used to serialize the pksim project into a SQLite database.
+-	Castle (currently version 5.0.1). Excellent inversion of control container implementation.
+-	NUnit (currently version 3.13.3). Simple yet powerful unit test framework.
+-   Npoi (currently version 2.5.5). Used to read from and write to xml files. 
+-	FakeItEasy (currently version 7.3.1). Used in conjunction with NUnit to mock objects used in tests.
+-	Microsoft Extensions Logging (currently version 3.1.0). Used to log any warnings or error to a log file or to the console when debugging.
+
+
+## Architecture
+
+The architecture used in PKSim IS NOT based on the traditional layered infrastructure. The traditional layer creates per construction a tight coupling between layers, as each layers depends on the layer beneath it and in general all layers depends on some cross-layer concerns such as security, logging etc usually defined in some kind of “infrastructure” layer. 
+ 
+![Traditional Layered Architecture (image from https://lorifpeterson.com/?p=64)](assets/images/Traditional_Layered_Architecture.png)
+
+Instead, an infrastructure pattern was used called “Onion architecture”, that at the time of design of the OSPSuite was quite new and has been gaining popularity ever since.
+
+![Onion Architecture (image from https://medium.com/expedia-group-tech/onion-architecture-deed8a554423)](assets/images/onion_architecture.png)
+ 
+The idea behind this pattern is fairly simple: In a nutshell, ALL CODE CAN DEPEND ON LAYERS MORE CENTRAL, BUT CODE CANNOT DEPEND ON LAYER FURTHER OUT. The coupling is directed to the center of the architecture.
+The domain modal is always the center of the architecture and thus has no dependencies whatsoever on other layers. In the layer Domain Services, we would typically find interfaces providing serialization behavior. The implementation however would be on the outside of the architecture (Infrastructure) as the serialization involved databases or xml manipulations that have no place in the core.
+ 
+![Onion Architecture (image from https://jeffreypalermo.com/2008/07/the-onion-architecture-part-2/)](assets/images/onion_architecture_details.png)
+
+
+In this example, the `IConferenceRepository` interface is defined as a DomainServices that can thus be accessed from all presenters of the application. The implementation however resides in Infrastructure.
+This architecture relies HEAVILY on the use of the [Dependency Inversion Principle](http://en.wikipedia.org/wiki/Dependency_inversion_principle) 
+The application core needs implementation of core interfaces, and if those implementing classes reside at the edges of the application, we need some mechanism for injecting that code at runtime so the application can do something useful.
```

Comment:
I think it would be good to have another document explaining IoC-Container usage in OSP Suite in some more detail (e.g. Register-classes etc.)

Reply:
I agree. Created a new issue for that (https://github.com/Open-Systems-Pharmacology/developer-docs/issues/20), since its implementation is not part of this PR.